design/plugin-versioning.md

```diff
@@ -0,0 +1,201 @@
+
+As new plugin versions are added, the plugin manager will be modified to retrieve the latest version of a plugin kind, adapting earlier versions of the plugin kind if possible.
+This is to allow the remainder of the Velero codebase to assume that it will always interact with the latest version of a plugin.
+The manager will check the registry for the plugin kind and name, and if the requested version is not found, it will attempt to fetch previous versions of the plugin kind and adapt it if possible.
+It will be up to the author of new plugin versions to determine whether a previous version of a plugin can be adapted to work with the interface of the new version.
+
+```
+// GetObjectStore returns a restartableObjectStore for name.
+func (m *manager) GetObjectStore(name string) (v2.ObjectStore, error) {
+    name = sanitizeName(name)
+
+    restartableProcess, err := m.getRestartableProcess(framework.PluginKindObjectStoreV2, name)
+    if err != nil {
+        // Check if plugin was not found
+        if errors.Is(err, pluginNotFoundError) {
+            // Try again but with previous version
+            restartableProcess, err := m.getRestartableProcess(framework.PluginKindObjectStore, name)
+            if err != nil {
+                // No v1 version found, return
+                return nil, err
+            }
+
+            r := newRestartableObjectStore(name, restartableProcess)
+            // Adapt v1 plugin to v2
+            return newAdaptedV1ObjectStore(r), nil
+        } else {
+            return nil, err
+        }
+    }
+    return newRestartableObjectStoreV2(name, restartableProcess), nil
+}
+```
+
+If the previous version is not available, or can not be adapted to the latest version, an error will be returned as is currently the case when a plugin implementation for a particular kind and provider can not be found.
+
+There are situations where it may be beneficial to check at the point where a plugin API call is made whether it implements a specific version of the API.
+This is something that can be addressed with future amendments to this design, however it does not seem to be necessary at this time.
+
+### Restartable Plugin Process
+
+As new versions of plugins are added, new restartable implementations of plugins will also need to be created.
+These are currently located within "pkg/plugin/clientmgmt" but will be rearranged to be grouped by kind and version like other plugin files.
+
+## Versioning Considerations
+
+It should be noted that if changes are being made to a plugin's API, it will only be necessary to bump the version once within a release cycle, regardless of how many changes are made within that cycle.
+This is because the changes will only impact consumers when they upgrade the versions of the Velero library that they import.
+Once a new version of Velero has been released however, any further changes will need to follow the process above and use a new version.
+
```

Comment:
Would this include a commitment to not introduce plugin version changes in z-stream bugfix releases? i.e. if a particular plugin released v2 in 1.7.0, then we would not release a v3 plugin in 1.7.1 -- that would have to wait for 1.8.0?

Reply:
Yes, I think so - I can't think of a situation at the moment that would warrant a new plugin version being introduced in a bugfix release. I'll add some more details in this section to make that clearer.